// @workspace/api-client-react/src/generated/api.ts
/**
 * Generated by orval v8.18.0 🍺
 * Do not edit manually.
 * Api
 * Meu Clube - Plataforma multiempresa para gestão de clubes e condomínios
 * OpenAPI spec version: 0.1.0
 */
import {
  useMutation,
  useQuery
} from '@tanstack/react-query';
import type {
  MutationFunction,
  QueryFunction,
  QueryKey,
  UseMutationOptions,
  UseMutationResult,
  UseQueryOptions,
  UseQueryResult
} from '@tanstack/react-query';

import type {
  AdminUser,
  AuthSession,
  ChangePasswordInput,
  Company,
  CompanyDashboard,
  CompanyInput,
  CompanyStatusUpdate,
  CompanyUpdate,
  CompanyUser,
  CompanyUserInput,
  CompanyUserUpdate,
  CreateAdminUserInput,
  HealthStatus,
  LoginInput,
  Me,
  Plan,
  PlanInput,
  Reservation,
  ReservationInput,
  ReservationStatusUpdate,
  ResetCompanyUserPassword200,
  ResetPasswordInput,
  SetResponsibleInput,
  Structure,
  StructureInput,
  StructureUpdate,
  Subscription,
  SubscriptionInput
} from './api.schemas';

import { customFetch } from '../custom-fetch';
import type { ErrorType , BodyType } from '../custom-fetch';

type AwaitedInput<T> = PromiseLike<T> | T;

      type Awaited<O> = O extends AwaitedInput<infer T> ? T : never;


type SecondParameter<T extends (...args: never) => unknown> = Parameters<T>[1];



const withQueryKey = <T extends object, K>(query: T, queryKey: K): T & { queryKey: K } => {
  const result = { queryKey } as T & { queryKey: K };
  for (const key of Object.keys(query)) {
    // The explicit queryKey always wins, matching the previous
    // `{ ...query, queryKey }` spread where it was set last.
    if (key === 'queryKey') continue;
    Object.defineProperty(result, key, {
      enumerable: true,
      configurable: true,
      get: () => (query as Record<string, unknown>)[key],
    });
  }
  return result;
};

export const getHealthCheckUrl = () => {




  return `/api/healthz`
}

/**
 * @summary Health check
 */
export const healthCheck = async ( options?: RequestInit): Promise<HealthStatus> => {

  return customFetch<HealthStatus>(getHealthCheckUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getHealthCheckQueryKey = () => {
    return [
    `/api/healthz`
    ] as const;
    }


export const getHealthCheckQueryOptions = <TData = Awaited<ReturnType<typeof healthCheck>>, TError = ErrorType<unknown>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof healthCheck>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getHealthCheckQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof healthCheck>>> = ({ signal }) => healthCheck({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof healthCheck>>, TError, TData> & { queryKey: QueryKey }
}

export type HealthCheckQueryResult = NonNullable<Awaited<ReturnType<typeof healthCheck>>>
export type HealthCheckQueryError = ErrorType<unknown>


/**
 * @summary Health check
 */

export function useHealthCheck<TData = Awaited<ReturnType<typeof healthCheck>>, TError = ErrorType<unknown>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof healthCheck>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getHealthCheckQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getLoginUrl = () => {




  return `/api/auth/login`
}

/**
 * @summary Login with phone and password
 */
export const login = async (loginInput: LoginInput, options?: RequestInit): Promise<AuthSession> => {

  return customFetch<AuthSession>(getLoginUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(loginInput)
  }
);}




export const getLoginMutationOptions = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof login>>, TError,{data: BodyType<LoginInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof login>>, TError,{data: BodyType<LoginInput>}, TContext> => {

const mutationKey = ['login'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof login>>, {data: BodyType<LoginInput>}> = (props) => {
          const {data} = props ?? {};

          return  login(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type LoginMutationResult = NonNullable<Awaited<ReturnType<typeof login>>>
    export type LoginMutationBody = BodyType<LoginInput>
    export type LoginMutationError = ErrorType<void>

    /**
 * @summary Login with phone and password
 */
export const useLogin = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof login>>, TError,{data: BodyType<LoginInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof login>>,
        TError,
        {data: BodyType<LoginInput>},
        TContext
      > => {
      return useMutation(getLoginMutationOptions(options));
    }

export const getChangePasswordUrl = () => {




  return `/api/auth/password`
}

/**
 * @summary Change current user's password
 */
export const changePassword = async (changePasswordInput: ChangePasswordInput, options?: RequestInit): Promise<void> => {

  return customFetch<void>(getChangePasswordUrl(),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(changePasswordInput)
  }
);}




export const getChangePasswordMutationOptions = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof changePassword>>, TError,{data: BodyType<ChangePasswordInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof changePassword>>, TError,{data: BodyType<ChangePasswordInput>}, TContext> => {

const mutationKey = ['changePassword'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof changePassword>>, {data: BodyType<ChangePasswordInput>}> = (props) => {
          const {data} = props ?? {};

          return  changePassword(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type ChangePasswordMutationResult = NonNullable<Awaited<ReturnType<typeof changePassword>>>
    export type ChangePasswordMutationBody = BodyType<ChangePasswordInput>
    export type ChangePasswordMutationError = ErrorType<void>

    /**
 * @summary Change current user's password
 */
export const useChangePassword = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof changePassword>>, TError,{data: BodyType<ChangePasswordInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof changePassword>>,
        TError,
        {data: BodyType<ChangePasswordInput>},
        TContext
      > => {
      return useMutation(getChangePasswordMutationOptions(options));
    }

export const getLogoutUrl = () => {




  return `/api/auth/logout`
}

/**
 * @summary Logout current session
 */
export const logout = async ( options?: RequestInit): Promise<void> => {

  return customFetch<void>(getLogoutUrl(),
  {
    ...options,
    method: 'POST'


  }
);}




export const getLogoutMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof logout>>, TError,void, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof logout>>, TError,void, TContext> => {

const mutationKey = ['logout'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof logout>>, void> = () => {


          return  logout(requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type LogoutMutationResult = NonNullable<Awaited<ReturnType<typeof logout>>>

    export type LogoutMutationError = ErrorType<unknown>

    /**
 * @summary Logout current session
 */
export const useLogout = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof logout>>, TError,void, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof logout>>,
        TError,
        void,
        TContext
      > => {
      return useMutation(getLogoutMutationOptions(options));
    }

export const getGetMeUrl = () => {




  return `/api/auth/me`
}

/**
 * @summary Get current authenticated user
 */
export const getMe = async ( options?: RequestInit): Promise<Me> => {

  return customFetch<Me>(getGetMeUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetMeQueryKey = () => {
    return [
    `/api/auth/me`
    ] as const;
    }


export const getGetMeQueryOptions = <TData = Awaited<ReturnType<typeof getMe>>, TError = ErrorType<void>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getMe>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetMeQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getMe>>> = ({ signal }) => getMe({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getMe>>, TError, TData> & { queryKey: QueryKey }
}

export type GetMeQueryResult = NonNullable<Awaited<ReturnType<typeof getMe>>>
export type GetMeQueryError = ErrorType<void>


/**
 * @summary Get current authenticated user
 */

export function useGetMe<TData = Awaited<ReturnType<typeof getMe>>, TError = ErrorType<void>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getMe>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetMeQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getListCompaniesUrl = () => {




  return `/api/companies`
}

/**
 * @summary List all companies (super admin)
 */
export const listCompanies = async ( options?: RequestInit): Promise<Company[]> => {

  return customFetch<Company[]>(getListCompaniesUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListCompaniesQueryKey = () => {
    return [
    `/api/companies`
    ] as const;
    }


export const getListCompaniesQueryOptions = <TData = Awaited<ReturnType<typeof listCompanies>>, TError = ErrorType<unknown>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listCompanies>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListCompaniesQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listCompanies>>> = ({ signal }) => listCompanies({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listCompanies>>, TError, TData> & { queryKey: QueryKey }
}

export type ListCompaniesQueryResult = NonNullable<Awaited<ReturnType<typeof listCompanies>>>
export type ListCompaniesQueryError = ErrorType<unknown>


/**
 * @summary List all companies (super admin)
 */

export function useListCompanies<TData = Awaited<ReturnType<typeof listCompanies>>, TError = ErrorType<unknown>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listCompanies>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListCompaniesQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateCompanyUrl = () => {




  return `/api/companies`
}

/**
 * @summary Create a new company
 */
export const createCompany = async (companyInput: CompanyInput, options?: RequestInit): Promise<Company> => {

  return customFetch<Company>(getCreateCompanyUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(companyInput)
  }
);}




export const getCreateCompanyMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createCompany>>, TError,{data: BodyType<CompanyInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createCompany>>, TError,{data: BodyType<CompanyInput>}, TContext> => {

const mutationKey = ['createCompany'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createCompany>>, {data: BodyType<CompanyInput>}> = (props) => {
          const {data} = props ?? {};

          return  createCompany(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateCompanyMutationResult = NonNullable<Awaited<ReturnType<typeof createCompany>>>
    export type CreateCompanyMutationBody = BodyType<CompanyInput>
    export type CreateCompanyMutationError = ErrorType<unknown>

    /**
 * @summary Create a new company
 */
export const useCreateCompany = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createCompany>>, TError,{data: BodyType<CompanyInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createCompany>>,
        TError,
        {data: BodyType<CompanyInput>},
        TContext
      > => {
      return useMutation(getCreateCompanyMutationOptions(options));
    }

export const getGetCompanyUrl = (companyId: number,) => {




  return `/api/companies/${companyId}`
}

/**
 * @summary Get company by ID
 */
export const getCompany = async (companyId: number, options?: RequestInit): Promise<Company> => {

  return customFetch<Company>(getGetCompanyUrl(companyId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetCompanyQueryKey = (companyId: number,) => {
    return [
    `/api/companies/${companyId}`
    ] as const;
    }


export const getGetCompanyQueryOptions = <TData = Awaited<ReturnType<typeof getCompany>>, TError = ErrorType<unknown>>(companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCompany>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetCompanyQueryKey(companyId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getCompany>>> = ({ signal }) => getCompany(companyId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getCompany>>, TError, TData> & { queryKey: QueryKey }
}

export type GetCompanyQueryResult = NonNullable<Awaited<ReturnType<typeof getCompany>>>
export type GetCompanyQueryError = ErrorType<unknown>


/**
 * @summary Get company by ID
 */

export function useGetCompany<TData = Awaited<ReturnType<typeof getCompany>>, TError = ErrorType<unknown>>(
 companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCompany>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetCompanyQueryOptions(companyId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getUpdateCompanyUrl = (companyId: number,) => {




  return `/api/companies/${companyId}`
}

/**
 * @summary Update company
 */
export const updateCompany = async (companyId: number,
    companyUpdate: CompanyUpdate, options?: RequestInit): Promise<Company> => {

  return customFetch<Company>(getUpdateCompanyUrl(companyId),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(companyUpdate)
  }
);}




export const getUpdateCompanyMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateCompany>>, TError,{companyId: number;data: BodyType<CompanyUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof updateCompany>>, TError,{companyId: number;data: BodyType<CompanyUpdate>}, TContext> => {

const mutationKey = ['updateCompany'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof updateCompany>>, {companyId: number;data: BodyType<CompanyUpdate>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  updateCompany(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type UpdateCompanyMutationResult = NonNullable<Awaited<ReturnType<typeof updateCompany>>>
    export type UpdateCompanyMutationBody = BodyType<CompanyUpdate>
    export type UpdateCompanyMutationError = ErrorType<unknown>

    /**
 * @summary Update company
 */
export const useUpdateCompany = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateCompany>>, TError,{companyId: number;data: BodyType<CompanyUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof updateCompany>>,
        TError,
        {companyId: number;data: BodyType<CompanyUpdate>},
        TContext
      > => {
      return useMutation(getUpdateCompanyMutationOptions(options));
    }

export const getUpdateCompanyStatusUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/status`
}

/**
 * @summary Change company status (activate, suspend, cancel)
 */
export const updateCompanyStatus = async (companyId: number,
    companyStatusUpdate: CompanyStatusUpdate, options?: RequestInit): Promise<Company> => {

  return customFetch<Company>(getUpdateCompanyStatusUrl(companyId),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(companyStatusUpdate)
  }
);}




export const getUpdateCompanyStatusMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateCompanyStatus>>, TError,{companyId: number;data: BodyType<CompanyStatusUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof updateCompanyStatus>>, TError,{companyId: number;data: BodyType<CompanyStatusUpdate>}, TContext> => {

const mutationKey = ['updateCompanyStatus'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof updateCompanyStatus>>, {companyId: number;data: BodyType<CompanyStatusUpdate>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  updateCompanyStatus(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type UpdateCompanyStatusMutationResult = NonNullable<Awaited<ReturnType<typeof updateCompanyStatus>>>
    export type UpdateCompanyStatusMutationBody = BodyType<CompanyStatusUpdate>
    export type UpdateCompanyStatusMutationError = ErrorType<unknown>

    /**
 * @summary Change company status (activate, suspend, cancel)
 */
export const useUpdateCompanyStatus = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateCompanyStatus>>, TError,{companyId: number;data: BodyType<CompanyStatusUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof updateCompanyStatus>>,
        TError,
        {companyId: number;data: BodyType<CompanyStatusUpdate>},
        TContext
      > => {
      return useMutation(getUpdateCompanyStatusMutationOptions(options));
    }

export const getListCompanyUsersUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/users`
}

/**
 * @summary List users linked to a company
 */
export const listCompanyUsers = async (companyId: number, options?: RequestInit): Promise<CompanyUser[]> => {

  return customFetch<CompanyUser[]>(getListCompanyUsersUrl(companyId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListCompanyUsersQueryKey = (companyId: number,) => {
    return [
    `/api/companies/${companyId}/users`
    ] as const;
    }


export const getListCompanyUsersQueryOptions = <TData = Awaited<ReturnType<typeof listCompanyUsers>>, TError = ErrorType<unknown>>(companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listCompanyUsers>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListCompanyUsersQueryKey(companyId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listCompanyUsers>>> = ({ signal }) => listCompanyUsers(companyId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listCompanyUsers>>, TError, TData> & { queryKey: QueryKey }
}

export type ListCompanyUsersQueryResult = NonNullable<Awaited<ReturnType<typeof listCompanyUsers>>>
export type ListCompanyUsersQueryError = ErrorType<unknown>


/**
 * @summary List users linked to a company
 */

export function useListCompanyUsers<TData = Awaited<ReturnType<typeof listCompanyUsers>>, TError = ErrorType<unknown>>(
 companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listCompanyUsers>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListCompanyUsersQueryOptions(companyId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getAddCompanyUserUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/users`
}

/**
 * @summary Add or invite a user to a company
 */
export const addCompanyUser = async (companyId: number,
    companyUserInput: CompanyUserInput, options?: RequestInit): Promise<CompanyUser> => {

  return customFetch<CompanyUser>(getAddCompanyUserUrl(companyId),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(companyUserInput)
  }
);}




export const getAddCompanyUserMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof addCompanyUser>>, TError,{companyId: number;data: BodyType<CompanyUserInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof addCompanyUser>>, TError,{companyId: number;data: BodyType<CompanyUserInput>}, TContext> => {

const mutationKey = ['addCompanyUser'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof addCompanyUser>>, {companyId: number;data: BodyType<CompanyUserInput>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  addCompanyUser(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type AddCompanyUserMutationResult = NonNullable<Awaited<ReturnType<typeof addCompanyUser>>>
    export type AddCompanyUserMutationBody = BodyType<CompanyUserInput>
    export type AddCompanyUserMutationError = ErrorType<unknown>

    /**
 * @summary Add or invite a user to a company
 */
export const useAddCompanyUser = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof addCompanyUser>>, TError,{companyId: number;data: BodyType<CompanyUserInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof addCompanyUser>>,
        TError,
        {companyId: number;data: BodyType<CompanyUserInput>},
        TContext
      > => {
      return useMutation(getAddCompanyUserMutationOptions(options));
    }

export const getResetCompanyUserPasswordUrl = (companyId: number,
    userId: number,) => {




  return `/api/companies/${companyId}/users/${userId}/reset-password`
}

/**
 * @summary Reset a company user's password (super admin only)
 */
export const resetCompanyUserPassword = async (companyId: number,
    userId: number,
    resetPasswordInput: ResetPasswordInput, options?: RequestInit): Promise<ResetCompanyUserPassword200> => {

  return customFetch<ResetCompanyUserPassword200>(getResetCompanyUserPasswordUrl(companyId,userId),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(resetPasswordInput)
  }
);}




export const getResetCompanyUserPasswordMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof resetCompanyUserPassword>>, TError,{companyId: number;userId: number;data: BodyType<ResetPasswordInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof resetCompanyUserPassword>>, TError,{companyId: number;userId: number;data: BodyType<ResetPasswordInput>}, TContext> => {

const mutationKey = ['resetCompanyUserPassword'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof resetCompanyUserPassword>>, {companyId: number;userId: number;data: BodyType<ResetPasswordInput>}> = (props) => {
          const {companyId,userId,data} = props ?? {};

          return  resetCompanyUserPassword(companyId,userId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type ResetCompanyUserPasswordMutationResult = NonNullable<Awaited<ReturnType<typeof resetCompanyUserPassword>>>
    export type ResetCompanyUserPasswordMutationBody = BodyType<ResetPasswordInput>
    export type ResetCompanyUserPasswordMutationError = ErrorType<unknown>

    /**
 * @summary Reset a company user's password (super admin only)
 */
export const useResetCompanyUserPassword = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof resetCompanyUserPassword>>, TError,{companyId: number;userId: number;data: BodyType<ResetPasswordInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof resetCompanyUserPassword>>,
        TError,
        {companyId: number;userId: number;data: BodyType<ResetPasswordInput>},
        TContext
      > => {
      return useMutation(getResetCompanyUserPasswordMutationOptions(options));
    }

export const getSetCompanyResponsibleUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/responsible`
}

/**
 * @summary Set the responsible user for a company
 */
export const setCompanyResponsible = async (companyId: number,
    setResponsibleInput: SetResponsibleInput, options?: RequestInit): Promise<Company> => {

  return customFetch<Company>(getSetCompanyResponsibleUrl(companyId),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(setResponsibleInput)
  }
);}




export const getSetCompanyResponsibleMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof setCompanyResponsible>>, TError,{companyId: number;data: BodyType<SetResponsibleInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof setCompanyResponsible>>, TError,{companyId: number;data: BodyType<SetResponsibleInput>}, TContext> => {

const mutationKey = ['setCompanyResponsible'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof setCompanyResponsible>>, {companyId: number;data: BodyType<SetResponsibleInput>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  setCompanyResponsible(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type SetCompanyResponsibleMutationResult = NonNullable<Awaited<ReturnType<typeof setCompanyResponsible>>>
    export type SetCompanyResponsibleMutationBody = BodyType<SetResponsibleInput>
    export type SetCompanyResponsibleMutationError = ErrorType<unknown>

    /**
 * @summary Set the responsible user for a company
 */
export const useSetCompanyResponsible = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof setCompanyResponsible>>, TError,{companyId: number;data: BodyType<SetResponsibleInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof setCompanyResponsible>>,
        TError,
        {companyId: number;data: BodyType<SetResponsibleInput>},
        TContext
      > => {
      return useMutation(getSetCompanyResponsibleMutationOptions(options));
    }

export const getUpdateCompanyUserUrl = (companyId: number,
    userId: number,) => {




  return `/api/companies/${companyId}/users/${userId}`
}

/**
 * @summary Update user role or status in a company
 */
export const updateCompanyUser = async (companyId: number,
    userId: number,
    companyUserUpdate: CompanyUserUpdate, options?: RequestInit): Promise<CompanyUser> => {

  return customFetch<CompanyUser>(getUpdateCompanyUserUrl(companyId,userId),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(companyUserUpdate)
  }
);}




export const getUpdateCompanyUserMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateCompanyUser>>, TError,{companyId: number;userId: number;data: BodyType<CompanyUserUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof updateCompanyUser>>, TError,{companyId: number;userId: number;data: BodyType<CompanyUserUpdate>}, TContext> => {

const mutationKey = ['updateCompanyUser'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof updateCompanyUser>>, {companyId: number;userId: number;data: BodyType<CompanyUserUpdate>}> = (props) => {
          const {companyId,userId,data} = props ?? {};

          return  updateCompanyUser(companyId,userId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type UpdateCompanyUserMutationResult = NonNullable<Awaited<ReturnType<typeof updateCompanyUser>>>
    export type UpdateCompanyUserMutationBody = BodyType<CompanyUserUpdate>
    export type UpdateCompanyUserMutationError = ErrorType<unknown>

    /**
 * @summary Update user role or status in a company
 */
export const useUpdateCompanyUser = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateCompanyUser>>, TError,{companyId: number;userId: number;data: BodyType<CompanyUserUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof updateCompanyUser>>,
        TError,
        {companyId: number;userId: number;data: BodyType<CompanyUserUpdate>},
        TContext
      > => {
      return useMutation(getUpdateCompanyUserMutationOptions(options));
    }

export const getRemoveCompanyUserUrl = (companyId: number,
    userId: number,) => {




  return `/api/companies/${companyId}/users/${userId}`
}

/**
 * @summary Remove user from a company
 */
export const removeCompanyUser = async (companyId: number,
    userId: number, options?: RequestInit): Promise<void> => {

  return customFetch<void>(getRemoveCompanyUserUrl(companyId,userId),
  {
    ...options,
    method: 'DELETE'


  }
);}




export const getRemoveCompanyUserMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof removeCompanyUser>>, TError,{companyId: number;userId: number}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof removeCompanyUser>>, TError,{companyId: number;userId: number}, TContext> => {

const mutationKey = ['removeCompanyUser'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof removeCompanyUser>>, {companyId: number;userId: number}> = (props) => {
          const {companyId,userId} = props ?? {};

          return  removeCompanyUser(companyId,userId,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type RemoveCompanyUserMutationResult = NonNullable<Awaited<ReturnType<typeof removeCompanyUser>>>

    export type RemoveCompanyUserMutationError = ErrorType<unknown>

    /**
 * @summary Remove user from a company
 */
export const useRemoveCompanyUser = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof removeCompanyUser>>, TError,{companyId: number;userId: number}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof removeCompanyUser>>,
        TError,
        {companyId: number;userId: number},
        TContext
      > => {
      return useMutation(getRemoveCompanyUserMutationOptions(options));
    }

export const getListStructuresUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/structures`
}

/**
 * @summary List bookable structures for a company
 */
export const listStructures = async (companyId: number, options?: RequestInit): Promise<Structure[]> => {

  return customFetch<Structure[]>(getListStructuresUrl(companyId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListStructuresQueryKey = (companyId: number,) => {
    return [
    `/api/companies/${companyId}/structures`
    ] as const;
    }


export const getListStructuresQueryOptions = <TData = Awaited<ReturnType<typeof listStructures>>, TError = ErrorType<unknown>>(companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listStructures>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListStructuresQueryKey(companyId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listStructures>>> = ({ signal }) => listStructures(companyId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listStructures>>, TError, TData> & { queryKey: QueryKey }
}

export type ListStructuresQueryResult = NonNullable<Awaited<ReturnType<typeof listStructures>>>
export type ListStructuresQueryError = ErrorType<unknown>


/**
 * @summary List bookable structures for a company
 */

export function useListStructures<TData = Awaited<ReturnType<typeof listStructures>>, TError = ErrorType<unknown>>(
 companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listStructures>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListStructuresQueryOptions(companyId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateStructureUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/structures`
}

/**
 * @summary Create a bookable structure
 */
export const createStructure = async (companyId: number,
    structureInput: StructureInput, options?: RequestInit): Promise<Structure> => {

  return customFetch<Structure>(getCreateStructureUrl(companyId),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(structureInput)
  }
);}




export const getCreateStructureMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createStructure>>, TError,{companyId: number;data: BodyType<StructureInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createStructure>>, TError,{companyId: number;data: BodyType<StructureInput>}, TContext> => {

const mutationKey = ['createStructure'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createStructure>>, {companyId: number;data: BodyType<StructureInput>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  createStructure(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateStructureMutationResult = NonNullable<Awaited<ReturnType<typeof createStructure>>>
    export type CreateStructureMutationBody = BodyType<StructureInput>
    export type CreateStructureMutationError = ErrorType<unknown>

    /**
 * @summary Create a bookable structure
 */
export const useCreateStructure = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createStructure>>, TError,{companyId: number;data: BodyType<StructureInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createStructure>>,
        TError,
        {companyId: number;data: BodyType<StructureInput>},
        TContext
      > => {
      return useMutation(getCreateStructureMutationOptions(options));
    }

export const getGetStructureUrl = (companyId: number,
    structureId: number,) => {




  return `/api/companies/${companyId}/structures/${structureId}`
}

/**
 * @summary Get a structure
 */
export const getStructure = async (companyId: number,
    structureId: number, options?: RequestInit): Promise<Structure> => {

  return customFetch<Structure>(getGetStructureUrl(companyId,structureId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetStructureQueryKey = (companyId: number,
    structureId: number,) => {
    return [
    `/api/companies/${companyId}/structures/${structureId}`
    ] as const;
    }


export const getGetStructureQueryOptions = <TData = Awaited<ReturnType<typeof getStructure>>, TError = ErrorType<unknown>>(companyId: number,
    structureId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getStructure>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetStructureQueryKey(companyId,structureId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getStructure>>> = ({ signal }) => getStructure(companyId,structureId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined && structureId !== null && structureId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getStructure>>, TError, TData> & { queryKey: QueryKey }
}

export type GetStructureQueryResult = NonNullable<Awaited<ReturnType<typeof getStructure>>>
export type GetStructureQueryError = ErrorType<unknown>


/**
 * @summary Get a structure
 */

export function useGetStructure<TData = Awaited<ReturnType<typeof getStructure>>, TError = ErrorType<unknown>>(
 companyId: number,
    structureId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getStructure>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetStructureQueryOptions(companyId,structureId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getUpdateStructureUrl = (companyId: number,
    structureId: number,) => {




  return `/api/companies/${companyId}/structures/${structureId}`
}

/**
 * @summary Update a structure
 */
export const updateStructure = async (companyId: number,
    structureId: number,
    structureUpdate: StructureUpdate, options?: RequestInit): Promise<Structure> => {

  return customFetch<Structure>(getUpdateStructureUrl(companyId,structureId),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(structureUpdate)
  }
);}




export const getUpdateStructureMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateStructure>>, TError,{companyId: number;structureId: number;data: BodyType<StructureUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof updateStructure>>, TError,{companyId: number;structureId: number;data: BodyType<StructureUpdate>}, TContext> => {

const mutationKey = ['updateStructure'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof updateStructure>>, {companyId: number;structureId: number;data: BodyType<StructureUpdate>}> = (props) => {
          const {companyId,structureId,data} = props ?? {};

          return  updateStructure(companyId,structureId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type UpdateStructureMutationResult = NonNullable<Awaited<ReturnType<typeof updateStructure>>>
    export type UpdateStructureMutationBody = BodyType<StructureUpdate>
    export type UpdateStructureMutationError = ErrorType<unknown>

    /**
 * @summary Update a structure
 */
export const useUpdateStructure = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateStructure>>, TError,{companyId: number;structureId: number;data: BodyType<StructureUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof updateStructure>>,
        TError,
        {companyId: number;structureId: number;data: BodyType<StructureUpdate>},
        TContext
      > => {
      return useMutation(getUpdateStructureMutationOptions(options));
    }

export const getDeleteStructureUrl = (companyId: number,
    structureId: number,) => {




  return `/api/companies/${companyId}/structures/${structureId}`
}

/**
 * @summary Remove a structure
 */
export const deleteStructure = async (companyId: number,
    structureId: number, options?: RequestInit): Promise<void> => {

  return customFetch<void>(getDeleteStructureUrl(companyId,structureId),
  {
    ...options,
    method: 'DELETE'


  }
);}




export const getDeleteStructureMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof deleteStructure>>, TError,{companyId: number;structureId: number}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof deleteStructure>>, TError,{companyId: number;structureId: number}, TContext> => {

const mutationKey = ['deleteStructure'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof deleteStructure>>, {companyId: number;structureId: number}> = (props) => {
          const {companyId,structureId} = props ?? {};

          return  deleteStructure(companyId,structureId,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type DeleteStructureMutationResult = NonNullable<Awaited<ReturnType<typeof deleteStructure>>>

    export type DeleteStructureMutationError = ErrorType<unknown>

    /**
 * @summary Remove a structure
 */
export const useDeleteStructure = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof deleteStructure>>, TError,{companyId: number;structureId: number}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof deleteStructure>>,
        TError,
        {companyId: number;structureId: number},
        TContext
      > => {
      return useMutation(getDeleteStructureMutationOptions(options));
    }

export const getListReservationsUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/reservations`
}

/**
 * @summary List reservations for a company
 */
export const listReservations = async (companyId: number, options?: RequestInit): Promise<Reservation[]> => {

  return customFetch<Reservation[]>(getListReservationsUrl(companyId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListReservationsQueryKey = (companyId: number,) => {
    return [
    `/api/companies/${companyId}/reservations`
    ] as const;
    }


export const getListReservationsQueryOptions = <TData = Awaited<ReturnType<typeof listReservations>>, TError = ErrorType<unknown>>(companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listReservations>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListReservationsQueryKey(companyId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listReservations>>> = ({ signal }) => listReservations(companyId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listReservations>>, TError, TData> & { queryKey: QueryKey }
}

export type ListReservationsQueryResult = NonNullable<Awaited<ReturnType<typeof listReservations>>>
export type ListReservationsQueryError = ErrorType<unknown>


/**
 * @summary List reservations for a company
 */

export function useListReservations<TData = Awaited<ReturnType<typeof listReservations>>, TError = ErrorType<unknown>>(
 companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listReservations>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListReservationsQueryOptions(companyId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateReservationUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/reservations`
}

/**
 * @summary Create a reservation
 */
export const createReservation = async (companyId: number,
    reservationInput: ReservationInput, options?: RequestInit): Promise<Reservation> => {

  return customFetch<Reservation>(getCreateReservationUrl(companyId),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(reservationInput)
  }
);}




export const getCreateReservationMutationOptions = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createReservation>>, TError,{companyId: number;data: BodyType<ReservationInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createReservation>>, TError,{companyId: number;data: BodyType<ReservationInput>}, TContext> => {

const mutationKey = ['createReservation'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createReservation>>, {companyId: number;data: BodyType<ReservationInput>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  createReservation(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateReservationMutationResult = NonNullable<Awaited<ReturnType<typeof createReservation>>>
    export type CreateReservationMutationBody = BodyType<ReservationInput>
    export type CreateReservationMutationError = ErrorType<void>

    /**
 * @summary Create a reservation
 */
export const useCreateReservation = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createReservation>>, TError,{companyId: number;data: BodyType<ReservationInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createReservation>>,
        TError,
        {companyId: number;data: BodyType<ReservationInput>},
        TContext
      > => {
      return useMutation(getCreateReservationMutationOptions(options));
    }

export const getGetReservationUrl = (companyId: number,
    reservationId: number,) => {




  return `/api/companies/${companyId}/reservations/${reservationId}`
}

/**
 * @summary Get reservation details
 */
export const getReservation = async (companyId: number,
    reservationId: number, options?: RequestInit): Promise<Reservation> => {

  return customFetch<Reservation>(getGetReservationUrl(companyId,reservationId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetReservationQueryKey = (companyId: number,
    reservationId: number,) => {
    return [
    `/api/companies/${companyId}/reservations/${reservationId}`
    ] as const;
    }


export const getGetReservationQueryOptions = <TData = Awaited<ReturnType<typeof getReservation>>, TError = ErrorType<unknown>>(companyId: number,
    reservationId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getReservation>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetReservationQueryKey(companyId,reservationId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getReservation>>> = ({ signal }) => getReservation(companyId,reservationId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined && reservationId !== null && reservationId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getReservation>>, TError, TData> & { queryKey: QueryKey }
}

export type GetReservationQueryResult = NonNullable<Awaited<ReturnType<typeof getReservation>>>
export type GetReservationQueryError = ErrorType<unknown>


/**
 * @summary Get reservation details
 */

export function useGetReservation<TData = Awaited<ReturnType<typeof getReservation>>, TError = ErrorType<unknown>>(
 companyId: number,
    reservationId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getReservation>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetReservationQueryOptions(companyId,reservationId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getUpdateReservationStatusUrl = (companyId: number,
    reservationId: number,) => {




  return `/api/companies/${companyId}/reservations/${reservationId}/status`
}

/**
 * @summary Confirm, cancel or reject a reservation
 */
export const updateReservationStatus = async (companyId: number,
    reservationId: number,
    reservationStatusUpdate: ReservationStatusUpdate, options?: RequestInit): Promise<Reservation> => {

  return customFetch<Reservation>(getUpdateReservationStatusUrl(companyId,reservationId),
  {
    ...options,
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(reservationStatusUpdate)
  }
);}




export const getUpdateReservationStatusMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateReservationStatus>>, TError,{companyId: number;reservationId: number;data: BodyType<ReservationStatusUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof updateReservationStatus>>, TError,{companyId: number;reservationId: number;data: BodyType<ReservationStatusUpdate>}, TContext> => {

const mutationKey = ['updateReservationStatus'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof updateReservationStatus>>, {companyId: number;reservationId: number;data: BodyType<ReservationStatusUpdate>}> = (props) => {
          const {companyId,reservationId,data} = props ?? {};

          return  updateReservationStatus(companyId,reservationId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type UpdateReservationStatusMutationResult = NonNullable<Awaited<ReturnType<typeof updateReservationStatus>>>
    export type UpdateReservationStatusMutationBody = BodyType<ReservationStatusUpdate>
    export type UpdateReservationStatusMutationError = ErrorType<unknown>

    /**
 * @summary Confirm, cancel or reject a reservation
 */
export const useUpdateReservationStatus = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof updateReservationStatus>>, TError,{companyId: number;reservationId: number;data: BodyType<ReservationStatusUpdate>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof updateReservationStatus>>,
        TError,
        {companyId: number;reservationId: number;data: BodyType<ReservationStatusUpdate>},
        TContext
      > => {
      return useMutation(getUpdateReservationStatusMutationOptions(options));
    }

export const getGetCompanyDashboardUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/dashboard`
}

/**
 * @summary Get dashboard summary for a company
 */
export const getCompanyDashboard = async (companyId: number, options?: RequestInit): Promise<CompanyDashboard> => {

  return customFetch<CompanyDashboard>(getGetCompanyDashboardUrl(companyId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetCompanyDashboardQueryKey = (companyId: number,) => {
    return [
    `/api/companies/${companyId}/dashboard`
    ] as const;
    }


export const getGetCompanyDashboardQueryOptions = <TData = Awaited<ReturnType<typeof getCompanyDashboard>>, TError = ErrorType<unknown>>(companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCompanyDashboard>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetCompanyDashboardQueryKey(companyId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getCompanyDashboard>>> = ({ signal }) => getCompanyDashboard(companyId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getCompanyDashboard>>, TError, TData> & { queryKey: QueryKey }
}

export type GetCompanyDashboardQueryResult = NonNullable<Awaited<ReturnType<typeof getCompanyDashboard>>>
export type GetCompanyDashboardQueryError = ErrorType<unknown>


/**
 * @summary Get dashboard summary for a company
 */

export function useGetCompanyDashboard<TData = Awaited<ReturnType<typeof getCompanyDashboard>>, TError = ErrorType<unknown>>(
 companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCompanyDashboard>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetCompanyDashboardQueryOptions(companyId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getListAdminUsersUrl = () => {




  return `/api/users`
}

/**
 * @summary List all super admin users
 */
export const listAdminUsers = async ( options?: RequestInit): Promise<AdminUser[]> => {

  return customFetch<AdminUser[]>(getListAdminUsersUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListAdminUsersQueryKey = () => {
    return [
    `/api/users`
    ] as const;
    }


export const getListAdminUsersQueryOptions = <TData = Awaited<ReturnType<typeof listAdminUsers>>, TError = ErrorType<unknown>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listAdminUsers>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListAdminUsersQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listAdminUsers>>> = ({ signal }) => listAdminUsers({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listAdminUsers>>, TError, TData> & { queryKey: QueryKey }
}

export type ListAdminUsersQueryResult = NonNullable<Awaited<ReturnType<typeof listAdminUsers>>>
export type ListAdminUsersQueryError = ErrorType<unknown>


/**
 * @summary List all super admin users
 */

export function useListAdminUsers<TData = Awaited<ReturnType<typeof listAdminUsers>>, TError = ErrorType<unknown>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listAdminUsers>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListAdminUsersQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateAdminUserUrl = () => {




  return `/api/users`
}

/**
 * @summary Create a new super admin user
 */
export const createAdminUser = async (createAdminUserInput: CreateAdminUserInput, options?: RequestInit): Promise<AdminUser> => {

  return customFetch<AdminUser>(getCreateAdminUserUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(createAdminUserInput)
  }
);}




export const getCreateAdminUserMutationOptions = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createAdminUser>>, TError,{data: BodyType<CreateAdminUserInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createAdminUser>>, TError,{data: BodyType<CreateAdminUserInput>}, TContext> => {

const mutationKey = ['createAdminUser'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createAdminUser>>, {data: BodyType<CreateAdminUserInput>}> = (props) => {
          const {data} = props ?? {};

          return  createAdminUser(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateAdminUserMutationResult = NonNullable<Awaited<ReturnType<typeof createAdminUser>>>
    export type CreateAdminUserMutationBody = BodyType<CreateAdminUserInput>
    export type CreateAdminUserMutationError = ErrorType<void>

    /**
 * @summary Create a new super admin user
 */
export const useCreateAdminUser = <TError = ErrorType<void>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createAdminUser>>, TError,{data: BodyType<CreateAdminUserInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createAdminUser>>,
        TError,
        {data: BodyType<CreateAdminUserInput>},
        TContext
      > => {
      return useMutation(getCreateAdminUserMutationOptions(options));
    }

export const getDeleteAdminUserUrl = (userId: number,) => {




  return `/api/users/${userId}`
}

/**
 * @summary Delete a super admin user
 */
export const deleteAdminUser = async (userId: number, options?: RequestInit): Promise<void> => {

  return customFetch<void>(getDeleteAdminUserUrl(userId),
  {
    ...options,
    method: 'DELETE'


  }
);}




export const getDeleteAdminUserMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof deleteAdminUser>>, TError,{userId: number}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof deleteAdminUser>>, TError,{userId: number}, TContext> => {

const mutationKey = ['deleteAdminUser'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof deleteAdminUser>>, {userId: number}> = (props) => {
          const {userId} = props ?? {};

          return  deleteAdminUser(userId,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type DeleteAdminUserMutationResult = NonNullable<Awaited<ReturnType<typeof deleteAdminUser>>>

    export type DeleteAdminUserMutationError = ErrorType<unknown>

    /**
 * @summary Delete a super admin user
 */
export const useDeleteAdminUser = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof deleteAdminUser>>, TError,{userId: number}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof deleteAdminUser>>,
        TError,
        {userId: number},
        TContext
      > => {
      return useMutation(getDeleteAdminUserMutationOptions(options));
    }

export const getListPlansUrl = () => {




  return `/api/plans`
}

/**
 * @summary List all commercial plans
 */
export const listPlans = async ( options?: RequestInit): Promise<Plan[]> => {

  return customFetch<Plan[]>(getListPlansUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListPlansQueryKey = () => {
    return [
    `/api/plans`
    ] as const;
    }


export const getListPlansQueryOptions = <TData = Awaited<ReturnType<typeof listPlans>>, TError = ErrorType<unknown>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listPlans>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListPlansQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listPlans>>> = ({ signal }) => listPlans({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listPlans>>, TError, TData> & { queryKey: QueryKey }
}

export type ListPlansQueryResult = NonNullable<Awaited<ReturnType<typeof listPlans>>>
export type ListPlansQueryError = ErrorType<unknown>


/**
 * @summary List all commercial plans
 */

export function useListPlans<TData = Awaited<ReturnType<typeof listPlans>>, TError = ErrorType<unknown>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listPlans>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListPlansQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreatePlanUrl = () => {




  return `/api/plans`
}

/**
 * @summary Create a commercial plan (super admin)
 */
export const createPlan = async (planInput: PlanInput, options?: RequestInit): Promise<Plan> => {

  return customFetch<Plan>(getCreatePlanUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(planInput)
  }
);}




export const getCreatePlanMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createPlan>>, TError,{data: BodyType<PlanInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createPlan>>, TError,{data: BodyType<PlanInput>}, TContext> => {

const mutationKey = ['createPlan'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createPlan>>, {data: BodyType<PlanInput>}> = (props) => {
          const {data} = props ?? {};

          return  createPlan(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreatePlanMutationResult = NonNullable<Awaited<ReturnType<typeof createPlan>>>
    export type CreatePlanMutationBody = BodyType<PlanInput>
    export type CreatePlanMutationError = ErrorType<unknown>

    /**
 * @summary Create a commercial plan (super admin)
 */
export const useCreatePlan = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createPlan>>, TError,{data: BodyType<PlanInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createPlan>>,
        TError,
        {data: BodyType<PlanInput>},
        TContext
      > => {
      return useMutation(getCreatePlanMutationOptions(options));
    }

export const getGetCompanySubscriptionUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/subscription`
}

/**
 * @summary Get subscription for a company
 */
export const getCompanySubscription = async (companyId: number, options?: RequestInit): Promise<Subscription> => {

  return customFetch<Subscription>(getGetCompanySubscriptionUrl(companyId),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetCompanySubscriptionQueryKey = (companyId: number,) => {
    return [
    `/api/companies/${companyId}/subscription`
    ] as const;
    }


export const getGetCompanySubscriptionQueryOptions = <TData = Awaited<ReturnType<typeof getCompanySubscription>>, TError = ErrorType<unknown>>(companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCompanySubscription>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetCompanySubscriptionQueryKey(companyId);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getCompanySubscription>>> = ({ signal }) => getCompanySubscription(companyId, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: companyId !== null && companyId !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getCompanySubscription>>, TError, TData> & { queryKey: QueryKey }
}

export type GetCompanySubscriptionQueryResult = NonNullable<Awaited<ReturnType<typeof getCompanySubscription>>>
export type GetCompanySubscriptionQueryError = ErrorType<unknown>


/**
 * @summary Get subscription for a company
 */

export function useGetCompanySubscription<TData = Awaited<ReturnType<typeof getCompanySubscription>>, TError = ErrorType<unknown>>(
 companyId: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCompanySubscription>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetCompanySubscriptionQueryOptions(companyId,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getSetCompanySubscriptionUrl = (companyId: number,) => {




  return `/api/companies/${companyId}/subscription`
}

/**
 * @summary Assign a plan to a company
 */
export const setCompanySubscription = async (companyId: number,
    subscriptionInput: SubscriptionInput, options?: RequestInit): Promise<Subscription> => {

  return customFetch<Subscription>(getSetCompanySubscriptionUrl(companyId),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(subscriptionInput)
  }
);}




export const getSetCompanySubscriptionMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof setCompanySubscription>>, TError,{companyId: number;data: BodyType<SubscriptionInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof setCompanySubscription>>, TError,{companyId: number;data: BodyType<SubscriptionInput>}, TContext> => {

const mutationKey = ['setCompanySubscription'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof setCompanySubscription>>, {companyId: number;data: BodyType<SubscriptionInput>}> = (props) => {
          const {companyId,data} = props ?? {};

          return  setCompanySubscription(companyId,data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type SetCompanySubscriptionMutationResult = NonNullable<Awaited<ReturnType<typeof setCompanySubscription>>>
    export type SetCompanySubscriptionMutationBody = BodyType<SubscriptionInput>
    export type SetCompanySubscriptionMutationError = ErrorType<unknown>

    /**
 * @summary Assign a plan to a company
 */
export const useSetCompanySubscription = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof setCompanySubscription>>, TError,{companyId: number;data: BodyType<SubscriptionInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof setCompanySubscription>>,
        TError,
        {companyId: number;data: BodyType<SubscriptionInput>},
        TContext
      > => {
      return useMutation(getSetCompanySubscriptionMutationOptions(options));
    }
