// @workspace/api-client-react/src/generated/api.schemas.ts
/**
 * Generated by orval v8.18.0 🍺
 * Do not edit manually.
 * Api
 * Meu Clube - Plataforma multiempresa para gestão de clubes e condomínios
 * OpenAPI spec version: 0.1.0
 */
export interface AdminUser {
  id: number;
  name: string;
  phone: string;
  email?: string | null;
  role: string;
  createdAt: string;
}

export interface CreateAdminUserInput {
  name: string;
  phone: string;
  email?: string;
  /** @minLength 6 */
  password: string;
}

export interface HealthStatus {
  status: string;
}

export interface LoginInput {
  phone: string;
  password: string;
}

export interface AuthSession {
  id: number;
  name: string;
  phone: string;
  /** Global role: super_admin or user */
  role: string;
  /**
     * Active company context
     * @nullable
     */
  companyId?: number | null;
}

export interface Me {
  id: number;
  name: string;
  phone: string;
  /** @nullable */
  email?: string | null;
  role: string;
  /** @nullable */
  companyId?: number | null;
  /**
     * Role within active company: admin, operator or member
     * @nullable
     */
  companyRole?: string | null;
}

export interface Company {
  id: number;
  name: string;
  /** club, condominium or other */
  type: string;
  /** active, inactive, suspended, cancelled */
  status: string;
  /** @nullable */
  email?: string | null;
  /** @nullable */
  phone?: string | null;
  /** @nullable */
  address?: string | null;
  createdAt: string;
  /** @nullable */
  planId?: number | null;
  /**
     * ID of the responsible user for this company
     * @nullable
     */
  responsibleUserId?: number | null;
}

export interface ChangePasswordInput {
  currentPassword: string;
  newPassword: string;
}

export interface ResetPasswordInput {
  /**
     * New password to set for the user
     * @minLength 6
     */
  newPassword: string;
}

export interface SetResponsibleInput {
  /** ID of the user to set as responsible */
  userId: number;
}

export interface CompanyInput {
  name: string;
  type: string;
  email?: string;
  phone?: string;
  address?: string;
  /** Name of the initial admin user */
  adminName?: string;
  /** Phone (login) of the initial admin user */
  adminPhone?: string;
  /** Password for the initial admin user */
  adminPassword?: string;
}

export interface CompanyUpdate {
  name?: string;
  email?: string;
  phone?: string;
  address?: string;
}

export interface CompanyStatusUpdate {
  /** active, inactive, suspended, cancelled */
  status: string;
}

export interface CompanyUser {
  id: number;
  userId: number;
  companyId: number;
  name: string;
  phone: string;
  /** @nullable */
  email?: string | null;
  /** admin, operator, member */
  role: string;
  /** active, inactive */
  status: string;
  createdAt?: string;
}

export interface CompanyUserInput {
  name: string;
  phone: string;
  email?: string;
  password?: string;
  /** admin, operator, member */
  role: string;
}

export interface CompanyUserUpdate {
  role?: string;
  status?: string;
}

export interface Structure {
  id: number;
  companyId: number;
  name: string;
  /** court, pool, room, field, gym, other */
  type: string;
  /** @nullable */
  description?: string | null;
  capacity: number;
  /** active, inactive, maintenance */
  status: string;
  /**
     * HH:MM format
     * @nullable
     */
  openTime?: string | null;
  /**
     * HH:MM format
     * @nullable
     */
  closeTime?: string | null;
  /** @nullable */
  maxDurationMinutes?: number | null;
  requiresApproval?: boolean;
  createdAt?: string;
}

export interface StructureInput {
  name: string;
  type: string;
  description?: string;
  capacity: number;
  openTime?: string;
  closeTime?: string;
  maxDurationMinutes?: number;
  requiresApproval?: boolean;
}

export interface StructureUpdate {
  name?: string;
  description?: string;
  capacity?: number;
  status?: string;
  openTime?: string;
  closeTime?: string;
  maxDurationMinutes?: number;
  requiresApproval?: boolean;
}

export interface Reservation {
  id: number;
  companyId: number;
  structureId: number;
  structureName?: string;
  userId: number;
  userName?: string;
  date: string;
  /** HH:MM */
  startTime: string;
  /** HH:MM */
  endTime: string;
  /** requested, confirmed, cancelled, rejected, completed */
  status: string;
  /** @nullable */
  notes?: string | null;
  createdAt: string;
}

export interface ReservationInput {
  structureId: number;
  date: string;
  startTime: string;
  endTime: string;
  notes?: string;
}

export interface ReservationStatusUpdate {
  /** confirmed, cancelled, rejected, completed */
  status: string;
  reason?: string;
}

export interface StatusCount {
  status: string;
  count: number;
}

export interface CompanyDashboard {
  totalUsers: number;
  totalStructures: number;
  totalReservations: number;
  reservationsByStatus: StatusCount[];
  upcomingReservations: Reservation[];
  reservationsThisWeek?: number;
  activeStructures?: number;
}

export interface Plan {
  id: number;
  name: string;
  /** @nullable */
  description?: string | null;
  maxUsers: number;
  maxStructures: number;
  price: number;
  /** monthly, annual */
  periodicity: string;
  active?: boolean;
  createdAt?: string;
}

export interface PlanInput {
  name: string;
  description?: string;
  maxUsers: number;
  maxStructures: number;
  price: number;
  periodicity: string;
}

export interface Subscription {
  id: number;
  companyId: number;
  planId: number;
  planName?: string;
  /** @nullable */
  planMaxUsers?: number | null;
  /** @nullable */
  planMaxStructures?: number | null;
  /** active, overdue, cancelled */
  status: string;
  startDate: string;
  /** @nullable */
  nextBillingDate?: string | null;
}

export interface SubscriptionInput {
  planId: number;
}

export type ResetCompanyUserPassword200 = {
  ok?: boolean;
};
